lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,189 @@
+    length_of_tags = overridden_tag_length(name, string_to_add)
+    if length_of_tags > remaining_length
+      enter_ignored_level
+      return
+    end
+
+    # Save the tag so we can push it on at the end
+    @closing_tags.push name unless single_tag_element? name
+
+    append_to_truncated_string string_to_add, length_of_tags
+  end
+
+  # This method is called when the parser encounters characters between tags
+  def characters(decoded_string)
+    if max_length_reached || ignore_mode?
+      @something_has_been_truncated = max_length_reached
+      return
+    end
+
+    # Use encoded length, so &gt; counts as 4 bytes, not 1 (which is what '>' would give)
+    encoded_string = @html_coder.encode(decoded_string, :named)
+    string_to_append = if encoded_string.bytesize > remaining_length
+      # This is the line which prevents HTML entities getting truncated - treat them as a single char
+      str = @html_coder.encode(truncate_string(decoded_string), :named)
+      str << tail if @something_has_been_truncated && remaining_length >= 0
+      str
+    else
+      encoded_string
+    end
+    append_to_truncated_string string_to_append, string_to_append.bytesize
+  end
+
+  # This method is called when the parser encounters an comment
+  def comment(string)
+    comment = comment_tag(string)
+    append_to_truncated_string(comment) if comment.bytesize <= remaining_length
+  end
+
+  # This method is called when the parser encounters cdata. In practice, this also
+  # gets called for this style of comment inside an element:
+  #
+  #   <style><!--
+  #     /* Font Definitions */
+  #     @font-face
+  #       {font-family:Wingdings;
+  #       panose-1:5 0 0 0 0 0 0 0 0 0;}
+  #   --></style>
+  #
+  def cdata_block(string)
+    append_to_truncated_string(string) if string.bytesize <= remaining_length
+  end
+
+  # This method is called when the parser encounters a closing tag
+  def end_element(name)
+    if ignore_mode?
+      exit_ignored_level
+      return
+    end
+
+    return if max_length_reached || ignorable_tag?(name) || ignore_mode?
+
+    unless single_tag_element? name
+      @closing_tags.pop
+      # Don't count the length when closing a tag - it was accommodated when
+      # the tag was opened
+      append_to_truncated_string closing_tag(name), 0
+    end
+  end
+
+  def end_document
+    @closing_tags.reverse.each { |name| append_to_truncated_string closing_tag name }
+  end
+
+  private
+
+  def opening_tag(name, attributes)
+    attributes_string = attributes_to_string attributes
+    if single_tag_element? name
+      "<#{name}#{attributes_string}/>"
+    else
+      "<#{name}#{attributes_string}>"
+    end
+  end
+
+  def comment_tag(comment)
+    "<!--#{comment}-->"
+  end
+
+  def closing_tag(name)
+    "</#{name}>"
+  end
+
+  def remaining_length
+    max_length - estimated_length_with_tail
+  end
+
+  def estimated_length_with_tail
+    @estimated_length + (@something_has_been_truncated ? tail.bytesize : 0)
```

Comment:
[This line](https://github.com/zendesk/abbreviato/pull/1/files#diff-3c291160de3d24bf025ee43f9ca67804R60) in particular seems more confusing because of that.. I would expect the `tail`'s length to already be accounted for in the earlier check against `remaining_length` so we can just add it if needed.